### PR TITLE
Add Curry function with tests

### PR DESCRIPTION
**Pull request by Paryz**

### CODE REVIEW

File: lib/fe.ex [hunk 1]
```diff
@@ -15,4 +15,23 @@ defmodule FE do
   """
   @spec const(a) :: (any -> a) when a: var
   def const(a), do: fn _ -> a end
+  
+  @doc """
+  Transforms a n-ary function in an unary one that accepts its first
+  argument and returns a function expecting the next one, successively
```
Comment (studzien) [inline lib/fe.ex:21]: Would you mind adding a few `doctest` examples here?

I admit I had some difficulty with understanding what the function does on the first reading of this description.

How about something along these lines?

```
Transforms a n-ary function f(x1, x2, ..., xn) to a unary function that accepts the f's first argument x1 and returns the unary function that accepts x2, and so on until xn.
The unary function that accepts xn returns the same value as f(x1, x2, ..., xn).
```

File: lib/fe.ex [hunk 2]
```diff
@@ -15,4 +15,23 @@ defmodule FE do
...
+  """
+  @spec curry((... -> any())) :: (any() -> any())
+  def curry(fun) do
+    case :erlang.fun_info(fun, :arity) do
+      {_, 0} -> raise ArgumentError, "0-ary functions cannot be curried"
```
Comment (studzien) [inline lib/fe.ex:26]: I'm wondering if it wouldn't be cleaner if we always returned a zero argument function from the last step always.

I can see the following pros:
- no special case and raising for zero argument functions in `curry`;
- more in line with languages that have currying built-in (laziness);
- we should be able to improve the vague spec of this function.

File: test/fe_test.exs [hunk 3]
```diff
@@ -20,4 +20,29 @@ defmodule FETest do
     assert FE.const(a_number).(a_tuple) == a_number
     assert FE.const(a_tuple).(some_binary) == a_tuple
   end
+
+  describe "curry/1" do
+    test "function with arity bigger than 0 is curried" do
```
Comment (studzien) [inline test/fe_test.exs:25]: Do you mind splitting these examples into separate test cases?

**Review by studzien — CHANGES_REQUESTED** [review]
Thanks!

A `curry` function design and a few cosmetic changes proposed

**Comment by tomekowal** [comment]
I don't want to be a buzzkill, but I am not sure if it is a good idea to add the `curry` function to Erlang. AFAIU anonymous functions add to call stack making it less readable and impact reduction count (5 arity function becomes 5 reductions instead of 1).

Currying is useful with Haskell point-free programming but in Erlang I'd rather use a function that partially applies first n arguments e.g. given `f(a,b,c,d,e)` `not_really_curry(f, 1, 2) -> g(c, d, e)`
Elixir dealt with that nicely with `&(1, 2, &1, &2, &3)` syntax.

I understand that currying is a useful technique in other programming languages, but it just doesn't fit the Erlang ecosystem as nicely as other solutions from this library.

**Pull request closed without merging**